File: frontend/lib/dailyLimit.js
/**
 * lib/dailyLimit.js
 * -----------------
 * Tiny localStorage-backed per-day quota for the PoC (fair-use limits like
 * "1 grievance / day" and "5 supports / day"). Resets automatically at
 * midnight (keyed on the local date string).
 */

function todayKey() {
  const d = new Date();
  return `${d.getFullYear()}-${d.getMonth() + 1}-${d.getDate()}`;
}

/** Current state: { used, remaining, max }. Safe during SSR. */
export function dailyState(key, max) {
  if (typeof window === "undefined") return { used: 0, remaining: max, max };
  try {
    const raw = JSON.parse(localStorage.getItem(key) || "{}");
    const used = raw.date === todayKey() ? raw.used || 0 : 0;
    return { used, remaining: Math.max(0, max - used), max };
  } catch {
    return { used: 0, remaining: max, max };
  }
}

/** Consume one unit and return the new state (no-op past the limit). */
export function consumeDaily(key, max) {
  const s = dailyState(key, max);
  if (s.remaining <= 0) return s;
  const used = s.used + 1;
  try { localStorage.setItem(key, JSON.stringify({ date: todayKey(), used })); } catch {}
  return { used, remaining: Math.max(0, max - used), max };
}

export const GRIEVANCE_LIMIT_KEY = "nk_grievance_quota";
export const SUPPORT_LIMIT_KEY = "nk_support_quota";

File: mobile/lib/domain/daily_limit.dart
import 'package:shared_preferences/shared_preferences.dart';

/// Port of lib/dailyLimit.js — tiny per-day fair-use quota ("1 grievance/day",
/// "5 supports/day"). Display-only in the demo: never blocks the action.
/// Resets automatically at midnight (keyed on the local date string).
const kGrievanceLimitKey = 'nk_grievance_quota';
const kSupportLimitKey = 'nk_support_quota';

class DailyState {
  const DailyState({required this.used, required this.remaining, required this.max});

  final int used;
  final int remaining;
  final int max;
}

class DailyLimit {
  DailyLimit(this._prefs, {DateTime Function()? now})
      : _now = now ?? DateTime.now;

  final SharedPreferences _prefs;
  final DateTime Function() _now;

  String _todayKey() {
    final d = _now();
    return '${d.year}-${d.month}-${d.day}';
  }

  DailyState state(String key, int max) {
    final date = _prefs.getString('$key.date');
    final used = date == _todayKey() ? (_prefs.getInt('$key.used') ?? 0) : 0;
    return DailyState(used: used, remaining: (max - used).clamp(0, max), max: max);
  }

  /// Consume one unit and return the new state (no-op past the limit).
  Future<DailyState> consume(String key, int max) async {
    final s = state(key, max);
    if (s.remaining <= 0) return s;
    final used = s.used + 1;
    await _prefs.setString('$key.date', _todayKey());
    await _prefs.setInt('$key.used', used);
    return DailyState(used: used, remaining: (max - used).clamp(0, max), max: max);
  }
}
